source/polygon.cpp

```diff
@@ -0,0 +1,46 @@
+ld perimeter(vector<Point<ll>>& poli){
+    ld res = 0.0;
+    int n = ssize(poli);
+    for(int i = 1; i <= n; ++i) res += poli[(i-1) % n].dist(poli[(i)%n]);
+    return res;
+}
+
+// returns 2*area; area < 0 if clock-wise, area > 0 if ccw
+ll double_area(vector<Point<ll>>& poli){
+    ll res = 0.0;
+    int n = ssize(poli);
+    for(int i = 0; i < n; ++i)
+        res += (poli[(i)%n].x * poli[(i+1)%n].y - poli[(i+1)%n].x*poli[(i)%n].y);
+    return res;
+}
+
+bool check_convex(vector<Point<ll>>& poli){
+    int n = ssize(poli);
+    if(n < 3) return false;
+    bool ft = poli[2].ccw_check(poli[0], poli[1]);
+    for(int i = 1; i < n-1; ++i)
+        if((poli[(i+2)%n]).ccw_check(poli[i%n], poli[(i+1)%n]) != ft){
+            return false; // concave
+        }
+    return true; // convex
+}
+
+// returns 1/0/-1 if pt inside/on vertex or edge/outside poli
+int point_inside_polygon(Point<ll> pt, vector<Point<ll>>& poli){
+    int n = ssize(poli);
+    if(n < 3) return -1;
```

Comment:
Another technicality - should we return 0 if the point is on the line (when given 2 points) / equal to the only given point? Unsure about this one

Reply:
Coming back to this one, since we're already going to use on_segment in the same function we may as well use it here too for the N=2 case

Reply:
fiiixed